lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,865 @@
+        if ((i < NumElements - 1 && FieldSizes[i] > FieldSizes[i + 1]) ||
+            (i == NumElements - 1 && FieldSizes[i] > FieldSizes[i - 1]))
+          // Padding can't be larger than the field before or after
+          FieldDI[i] = new FieldDebugInfo(FieldNum++);
+        else if ((i < NumElements - 1 && Address % FieldSizes[i + 1] != 0 &&
+                  (Address + FieldSizes[i]) % FieldSizes[i + 1] == 0) ||
+                 (i == NumElements - 1 && Address % FieldSizes[i - 1] != 0 &&
+                  (Address + FieldSizes[i]) % FieldSizes[i - 1] == 0)) {
+          // If remove this field, the next/previous field is not aligned and
+          // add this field, it is aligned, it's likely to be a padding
+          unsigned MaxWCP = 0;
+          dbgs() << "Considering F" << i + 1 << "\n";
+          for (unsigned j = 0; j < NumElements; j++) {
+            auto *Pair = CPBuilder->getCloseProximityPair(i, j);
+            dbgs() << "(" << Pair->first << "," << Pair->second << ")\n";
+            if (Pair->first > MaxWCP)
+              MaxWCP = Pair->first;
+          }
+          if (MaxWCP != 0)
+            FieldDI[i] = new FieldDebugInfo(FieldNum++);
+          else
+            // This is highly likely to be a padding because there's no WCP with
+            // other fields
+            FieldDI[i] = nullptr;
+        } else {
+          FieldDI[i] = new FieldDebugInfo(FieldNum++);
+        }
+        Address += FieldSizes[i];
+      }
+    }
+  }
+}
+
+// Functions of FieldReorderAnalyzer
+// Constructor of FieldReorderAnalyzer. Need to calculate CP relations
+// according to the new reordering algorithm and based on CloseProximityBuilder
+// passed in. Constructor also selects starting fields to put in to the NewOrder
+// that is going to be a recommended ordering. Other variables are initialized
+// in the base class constructor
+FieldReorderAnalyzer::FieldReorderAnalyzer(const Module &CM,
+                                           const StructType *ST,
+                                           const CloseProximityBuilder *CPB,
+                                           const DICompositeType *DI)
+    : StructTransformAnalyzer(CM, ST, CPB, DI) {
+  DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                  dbgs() << "Initializing FieldReorderAnalyzer of size: "
+                         << NumElements << "\n");
+  // Calculate every CP relations and save it to an array to avoid repeated
+  // calculation
+  CloseProximityRelations.resize(NumElements);
+  double MaxWCP = 0;
+  FieldNumType MaxI, MaxJ;
+  for (unsigned i = 0; i < NumElements; i++) {
+    CloseProximityRelations[i].resize(NumElements);
+    for (unsigned j = 0; j < NumElements; j++) {
+      CloseProximityRelations[i][j] = calculateCloseProximity(i, j);
+      if (i < j && CloseProximityRelations[i][j] > MaxWCP) {
+        MaxWCP = CloseProximityRelations[i][j];
+        MaxI = i;
+        MaxJ = j;
+      }
+    }
+  }
+  if (MaxWCP < 1e-3) {
```

Comment:
I mentioned in the previous review. This should probably be a function (almostZero or smth like that) common to all your code